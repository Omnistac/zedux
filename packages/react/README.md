# Atomic Architecture

## What is an atom?

An atom can be:

- a simple useState-esque state container
- a selector, deriving state from other atoms
- a function that runs an async flow/side effect
- a wrapper around a Zedux store

the factory can return:

- a normal value - will be put in a store somehow
- an observable - will be subscribed to and [TODO: the latest emission? An accumulation of all emissions?] will be stored in a store somehow
- a store
- a promise - the promise will be awaited and an internal state machine will advance to success or failure state (TODO: are retries possible?) on resolve/reject. The Promise result should be one of the three above types

when you `useAtom` an atom, it is instantiated if it hasn't been.

## does the factory function run every time a dependency updates?

no. Factory runs exactly once per atom instance.

## so .. how would an atom be notified of a dependency's updates?

Hmm.. Could make it possible to import streams. .. But then you'd be dealing with multiple subscriptions

Zedux stores solve this! If all atoms export a store, all atoms can be injected anywhere - in React or in other atoms or literally anywhere - using the same subscription model.

## how would you combine state from 2 atoms?

```ts
const addAtom = createAtom({
  key: 'add',
  factory: () => {
    const counterAtom1 = inject(counterAtom, [1])
    const counterAtom2 = inject(counterAtom, [2])

    counterAtom1.subscribe() // ..... no. Don't want to handle subscriptions manually here.

    // hmm this suffers from the same multiple-subscribers problem:
    return combineLatest(
      counterAtom1.getState$(),
      counterAtom2.getState$()
    ).pipe(map(([a, b]) => a + b))
  },
})
```

so stores it is

```ts
const addAtom = createAtom({
  key: 'add',
  factory: () => {
    const counterStore1 = inject(counterAtom, [1])
    const counterStore2 = inject(counterAtom, [2])

    return combineLatest(from(counterStore1), from(counterStore2)).pipe(
      map(([a, b]) => a + b)
    )
  },
})
```

## what do local atoms look like?

```tsx
const Child = () => {}

const Parent = () => {
  const localAtom = someLocalAtom.useAtom() // is some id required?

  return (
    <LocalProvider atom={localAtom}>
      <Child />
    </LocalProvider>
  )
}
```

## TODO: how would you subscribe, receive all past events, and then continue listening to future ones

## TODO: how does cleanup work? If e.g. a factory function has subscriptions to clean up when its atom instance is destroyed.

could be solved with Zedux stores - e.g. a store could have a `destroy` method that triggers a `effectType: DESTROY` effect that can be hooked into.

## can the factory run in a react context?

no. Def no for app or global atoms. Local atoms could. Would a special local-atoms-only api make sense for that?

## TODO: timeout for async atoms - throw error for React error boundaries to catch.

## Zedux todos:

- fix createActor types (DONE)
- add createStore overload with initialState
- remove { error } from dispatch result
- fix Subscriber overloads
- emit current state on observablified stores
- disallow extra keys in returned state in `.reduce()`

# Internal Structure

## `globalStore`

This export holds all the state Zedux needs to function. Do not modify this directly. DevTool components can hook into this (read-only!) to provide information about the whole Zedux ecosystem for debugging and dev X features.

Its structure is:

```ts
globalStoreState = {
  atoms: {
    'some-global-atom-key': {
      implementations: {
        // 'im' stands for 'implementation' - these are internally generated ids
        'im-0': GlobalAtom,
        'im-1': GlobalAtom // different implementations should not change the scope (to GlobalAtom or LocalAtom in this case) - will log error
      },
      instances: {
        'in-0': {
          addSubscriber: (subscriber: Subscriber<T>) => () => void, // TODO: maybe Zedux could propagate an effect on subscribe
          dependencies: [],
          key: 'some-global-atom-key',
          metaStore: Store<AtomMetadata<T>>,
          stateStore?: Store<T>,
          ttl?: number | Observable<any>,
        },
        'in-1': AtomInstance
      },
      overrides: {
        // 'ov' stands for 'override' - these are internally generated ids
        'ov-0': GlobalAtom // scope shouldn't change here either
      }
    },
    'some-local-atom-key': {
      implementations: { ... },
      instances: {
        'in-0': AtomInstance
      }
    }
  },
  molecules: {
    // molecules can't be overridden (DI) or have multiple implementations or instantiations
    'some-molecule-key': Molecule,
    'another-molecule-key': Molecule
  },
  pools: {
    global: {
      instances: {
        'some-global-atom-key': {
          _default: 'in-0',
          '{some:"param"}': 'in-1'
        },
        'some-local-atom-key': {}
      }
    },
    'app-0': {
      instances: {
        'some-app-atom-key': {
          _default: 'in-0'
        }
      },
      overrides: {
        'some-app-atom-key': 'ov-0'
      }
    }
  }
}
```

Fully qualified ids for referring to atom implementations, instances, and overrides are tuples like `['some-atom-key', 'ai-0']`

### factory hooks (injectors?)

```tsx
const todosAtom = atom('todos', [])

const filteredTodosAtom = atom({
  key: 'filteredTodos',
  value: (isDone: boolean) => {
    const todos = todosAtom.injectValue()

    return todos.filter(todo => todo.isDone === isDone)
  },
})

const Filter = () => {
  const finishedTodos = filteredTodosAtom.useValue(true)
  const unfinishedTodos = filteredTodosAtom.useValue(false)

  return (
    <>
      <h3>finished todos:</h3>
      <ul>
        {finishedTodos.map(todo => (
          <li>{todo.text}</li>
        ))}
      </ul>
      <h3>unfinished todos:</h3>
      <ul>{unfinishedTodos.map()}</ul>
    </>
  )
}
```

async example

```tsx
import axios from 'axios'

const axiosAtom = atom({
  key: 'axios',
  readonly: true,
  value: axios
})

const usersListAtom = atom('usersList', () => {
  const axios = axiosAtom.injectValue()
  const filters = filtersAtom.injectValue()
  const store = injectConstant(() => createStore().hydrate([]))

  injectEffect(() => {
    const { data } = await axios.post('http://localhost/users/fetch', {
      filters,
    })
    store.setState(data)
  }, [axios, filters])

  return store
})

// SomeComponent
const usersList = usersListAtom.useValue()


const electronNetwork = axiosAtom.override(() => ({
  post: (...args) => network.request(...args)
}))

// ElectronApp
<AppProvider atoms={[electronNetwork]}>
  <SomeComponent />
</AppProvider>
```

streams example

```tsx
const usersStreamAtom = atom({
  key: 'usersStream',
  readonly: true,
  value: () => {
    const openSocket = openSocketAtom.useValue()

    return from(openSocket('wss://localhost/users'))
  },
})

const usersReducer = createReducer([])
// ... handle add, delete, update ...

const usersListAtom = atom('usersList', () => {
  const [users$, setUsers$] = usersStreamAtom.injectState()
  const store = injectConstant(() => createStore(usersReducer))

  injectEffect(() => {
    const subscription = users$.subscribe(store.dispatch)

    return subscription.unsubscribe
  }, [users$])

  return store
})

// some component

const [users, setUsers] = usersListAtom.useState()
```

```tsx
const filteredTodosStreamAtom = atom({
  key: 'filteredTodosStream',
  readonly: true,
  value: () => {
    const todos$ = todosAtom.injectValue$()
    const filters$ = filtersAtom.injectValue$()

    // injectConstant isn't needed in this particular atom 'cause nothing could cause it to recompute
    // TODO: That's actually a feature we could add - some `static: true` atom config option to turn off reactivity
    // (actually probably log an error if a reactive injector is used in such an atom)
    return injectConstant(
      () => todos$.pipe(
        withLatestFrom(filters$),
        filter(([todo, filters]) => todo.isDone !== filters.isDone),
        map(([todo]) => todo)
      )
    )
  }
})

// ... in some component:

const todos$ = filteredTodosStreamAtom.useValue()
```

## Current API

- `atom.useAction$()` - probably replace with a `useAction$(atom)` helper from @zedux/streams
- `atom.useApi()` - I do not know what this was supposed to be. Everything? Isn't that what `useInstance()` is for?
- `atom.useCallback()` - now `useMethods()`
- `atom.useDispatch()`
- `atom.useInstance()`
- `atom.useInvalidate()`
- `atom.useMetadata()` - ? Maybe. Not doing for now
- `atom.useSelector()`
- `atom.useSetState()`
- `atom.useState()`
- `atom.useStore()`
- `atom.useValue()`
- `atom.useValue$()` - probably replace with a `useValue$(atom)` helper from @zedux/streams

- `atom.useLocal()` .... no. This is just `localAtom.useValue()` or `*State()` or `*Dispatch()` or `*Store()` or `*Methods()` or `*Instance()`. No, correction, only `atom.useInstance()` and `atom.injectInstance()` do this. The other hooks are now an alias for `atom.useConsumer(...args).use*()` and same for injectors (obv..).
- `atom.useConsumer()`

- `atom.injectAction$()` - probably replace with a `injectAction$(atom)` helper from @zedux/streams
- `atom.injectApi()`
- `atom.injectCallback()`
- `atom.injectDispatch()`
- `atom.injectInstance()`
- `atom.injectInvalidate()`
- `atom.injectMetadata()` - ? Maybe. Not doing for now
- `atom.injectSelector()`
- `atom.injectSetState()`
- `atom.injectState()`
- `atom.injectStore()`
- `atom.injectValue()`
- `atom.injectValue$()` - probably replace with a `injectValue$(atom)` helper from @zedux/streams

- `atom.injectLocal()` ??? No. This is just `localAtom.injectValue()` or `*State()` or `*Dispatch()` or `*Store()` or `*Methods()` or `*Instance()`. Actually nope, see above correction for `atom.useLocal()`.
- `atom.injectConsumer()` ?? Nope. Wellll it is possible. `atom.use*` would have to `useContext(universalInstancesContext)` and pass that tree of provided instances to the atom diContext layer. Um except no, 'cause a given atom instance could be created outside this React component tree but still inside the AppProvider. So this has to come back to the AppProvider layer. Which isn't useful for differentiating instances in the same AppProvider. So no.


- `selector.useValue()`
- `selector.useValue$()`
- `selector.useMetadata()`
- `selector.useApi()` ?

- `selector.injectValue()`
- `selector.injectValue$()`
- `selector.injectMetadata()`
- `selector.injectApi()` ?


- `molecule.useValue()`
- `molecule.useValue$()`

- `molecule.injectValue()`
- `molecule.injectValue$()`


- `atom()`
- `molecule()`
- `selector()`
- `<AppProvider />`
- `<AtomInstanceProvider />`


- `injectCallback()`
- `injectConstant()`
- `injectEffect()`
- `injectMemo()`
- `injectRef()`


- `useGlobalStore()`


### `atom.useInstance()` and `atom.injectInstance()`

This hook/injector can be used to grab other hooks without re-passing params or recreating a subscription to the atom for performance reasons. Can also be used to pass an atom instance down to children components via context so they don't have to know/specify params either:

```tsx
const RegistrationForm = ({ children }) => {
  const formAtomInstance = formAtom.useInstance('registration-form')
  const form = formAtomInstance.useValue()
  const dispatch = formAtomInstance.useDispatch()

  return (
    <formAtomInstance.Provider>
      {children}
    </formAtomInstance.Provider>
  )
}

const Input = ({ name }) => {
  const formAtomInstance = formAtom.useConsumer() // TODO: we could make some way to notify formAtom of this consumer, passing params to initialize this form field
  const [form, setForm] = formAtomInstance.useState()

  return (
    <input
      name={name}
      onChange={({ target }) => setForm({ [name]: target.value })}
      value={value}
    />
  )
}

const Parent = () => (
  <RegistrationForm>
    <Input name="email" />
    <Input name="password" />
  </RegistrationForm>
)
```

## MOaR StuFf

NOTE: local atoms could take non-serializable params (including atom instances)

NOTE: could add a suspense preload hook thing

```tsx
function SomeComponent() {
  const loadSomeAtom = someAtom.useLazy() // doesn't actually create an instance

  return (
    <button onClick={() => {
      loadSomeAtom(params) // creates the instance
    }}>
      Click to load!
    </button>
  )
}
```

NOTE: could add more suspense stuff

```tsx
const someAtom = atom('some', () => {
  injectSuspender(async () => {
    const val = await someAsyncOperation()
    return val
  })
})

// or
const someAtom = atom('some', () => {
  const [machine, promise] = injectAsync(async machine => {
    machine.start()

    try {
      const result = await fetchSomething()
      machine.success(await result.json())
    } catch (err) {
      machine.error(err)
    }
  }, [])

  injectSuspender(promise)

  return machine
})
```


## Main Features Zedux Has That Other Libs Don't

- action streams
- isolated stores work well with code splitting
- designed to work with streams - e.g. reusing a single socket connection or managing lifecycles of socket.io, kafka, or gRPC data streams.

And it adds:

- DI
- state machines

## React Query Similarities

```tsx
import { injectQuery, useAtom } from '@zedux/react'

function Todos() {
  const { data, error, isFetching, status } = useAtom('todos', () => injectQuery(fetchTodos))
}

// dependent queries
const { isIdle, data: projects } = useQuery('projects', async () => {
  const userId = await currentUserIdAtom.injectValueAsync()

  return getProjectsByUser(userId)
})
```

## AtomContext example

```tsx
const reduxAtomContext = atomContext(INITIAL_REDUX_STATE)

function App() {
  const reduxState = useSelector(state => state)
  const reduxContext = reduxAtomContext.useInstance(reduxState)

  return (
    <AppProvider contexts={[reduxContext]}>
      <Routes />
    </AppProvider>
  )
}

const someAtom = atom('some', () => {
  const reduxInstance = reduxAtomContext.injectConsumer() // there is no injectInstance()
  const reduxState = reduxAtomContext.injectValue()
})

function SomeComponent() {
  const reduxInstance = reduxAtomContext.useConsumer() // there is a useInstance() - need to clearly document the distinction
  const reduxState = reduxAtomContext.useValue()
}
```

## Work List

Round 1 (end of December, 2020)

x- Make use* and inject* apis similar
x- Determine if factories are ok the way they are or if they should mimic hooks - rerendering on dep change
  x- (After Way Too MUCH debate, it's been decided to switch to a hooks/react-like approach)
x- Add `readonly` atoms. Good for top-level wrapper atoms (e.g. an `axios` atom) and selector atoms
x- Add `preload: init => init(params)` and `<AppProvider preload={() => { someAtom.useState(params) }}>`
  x- also make `preload` return `{children}` so they can be wrapped in `<AtomInstanceProvider instances={[...instances]}>`
x- Add a param to `store.getRefCount()` to filter out internal Zedux subscriptions
x- Expand `testMode` to a `flags` param and `isTestSafe` to a `flags` prop.
- Finish examples
- Write tests
- Write documentation

Round 2 (Feb 17, 2021)

x- context (`<AppProvider context={...}>` and `injectContext()`)
  x- now ~`contextStore()` (?)~ `atomContext()`
- `molecule()` and add molecule instantiation to atom instantiation
- remove duplicated code in `selector()` and `atom()`
- clean up types (does each atom, atom instance, and atom config really need its own interface)
- `atom.useDispatch()` and `atom.injectDispatch()`
- `atom.useState()` and `atom.injectState()`
- `atom.useSetState()` and `atom.injectSetState()`
x- `instance.Provider` and `atom.useConsumer()`
x- `injectInvalidate()`, `atom.injectInvalidate()`, `atom.useInvalidate()`
- `useStore()`, `atom.useStore()`, `atom.injectStore()`
x- `injectCallback()`
- all the standard atom factories - `globalAtom()`, `globalSelector()`, `localAtom()`, `localSelector()`
- all the specialized atom factories - `molecule()`, `query()`, `stream()`, `mutation()`
- all the inline hook factories? - `useGlobalAtom()`, `useGlobalSelector()`, `useLocalAtom()`, `useLocalSelector()`
- React Query stuff
  - add `atom.invalidateAll()`? No, it would have to use a hook. Maybe `const invalidate = atom.useInvalidate(); invalidate.all()`
  - `query()`, `mutation()`, `stream()` OR `promise()`, `stream()` OR `unary()`, `clientStream()`, `serverStream()`, `multiplexStream()` OR injectors - `injectQuery()` &c. or `injectPromise()` &c. or `injectUnary()` &c.
- should `useInstance()` everywhere be changed to `useProvider()`?
- `atom.useSelector()` and `atom.injectSelector()`
- a way to dynamically instantiate and access atoms and atomContexts - not using hooks (directly). Like React Query's `useQueries()` hook - https://react-query.tanstack.com/guides/parallel-queries#dynamic-parallel-queries-with-usequeries. I'm thinking at least `useAtomContexts()` and `useAtoms()` (and equivalent injectors).
- make AppProviders composable. Overrides and contexts from multiple parent AppProviders are merged together. Global atoms get added to the lowest pool that overrides them.
x- `injectWhy()`
- `methods` -> `exports`
- `maxInstances` - complements ttl. Use a FIFO queue. No instances will ever be cleaned up while in use. And none will be cleaned up while queueSize <= maxInstances. Stale instances will be scheduled for clean up when queueSize > maxInstances. Newly stale instances will be immediately scheduled for clean up if queueSize > maxInstances.
- should `useSelector()`/`injectSelector()` be changed to `useDerivation()`/`injectDerivation()`? The term "selector" is too overloaded.