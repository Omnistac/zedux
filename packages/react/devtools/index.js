if (process.env.NODE_ENV !== 'development') {
  module.exports = {
    getStateHub: function getStateHub() {},
    StateHub: function StateHub() {
      return null
    },
  }
} else {
  module.exports = require('./dev')
}
