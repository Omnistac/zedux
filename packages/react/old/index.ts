export * from './AppProvider'
export { globalStore } from './global/store'
export * from './createAtom'
export * from './createMolecule'
export * from './types'
export * from './useAtom'
export * from './useMolecule'
export * from './useStore'
