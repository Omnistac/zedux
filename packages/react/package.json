{
  "name": "@zedux/react",
  "version": "0.3.4",
  "description": "Elite state management for React",
  "main": "./dist/es/react/src/index.js",
  "module": "./dist/es/react/src/index.js",
  "es2015": "./dist/es/react/src/index.js",
  "publishConfig": {
    "registry": "https://omnistac.jfrog.io/omnistac/api/npm/npm-local/"
  },
  "typings": "./dist/es/react/src/index.d.ts",
  "sideEffects": false,
  "files": [
    "dist"
  ],
  "scripts": {
    "build-storybook": "build-storybook",
    "build:es": "ttsc --outDir dist/es -p tsconfig.es.json",
    "build:umd:min": "cross-env NODE_ENV=production webpack",
    "build:umd": "cross-env NODE_ENV=development webpack",
    "build": "yarn build:es && yarn build:umd && yarn build:umd:min",
    "dev": "yarn build:es -- --watch",
    "prepublishOnly": "yarn build",
    "storybook": "start-storybook -p 6006",
    "test": "cd ../.. && yarn test packages/react"
  },
  "keywords": [
    "abstraction",
    "atom",
    "atoms",
    "atomic",
    "beautiful",
    "boilerplate",
    "DI",
    "dependency",
    "easy",
    "flux",
    "functional",
    "immutable",
    "injection",
    "machine",
    "management",
    "memoized",
    "performant",
    "react",
    "reducer",
    "redux",
    "simple",
    "state",
    "test",
    "tests",
    "testing",
    "zedux"
  ],
  "author": "Joshua Claunch",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/Omnistac/zedux.git"
  },
  "bugs": {
    "url": "https://github.com/Omnistac/zedux/issues"
  },
  "homepage": "https://github.com/Omnistac/zedux#readme",
  "devDependencies": {
    "@testing-library/jest-dom": "^5.11.9",
    "@testing-library/react": "^11.2.5",
    "@types/styled-components": "^5.1.7",
    "core-js": "3",
    "react": "^17.0.1",
    "react-dom": "^17.0.1",
    "react-icons": "^4.1.0",
    "styled-components": "^5.2.1",
    "ttypescript": "^1.5.12",
    "typescript-transform-paths": "^2.2.3"
  },
  "dependencies": {},
  "peerDependencies": {
    "react": ">=17.0.1",
    "react-dom": ">=17.0.1"
  }
}
