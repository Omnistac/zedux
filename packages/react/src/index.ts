export * from '@zedux/atoms'
export * from './components/index'
export * from './hooks/index'
export * from './inject'
export * from './types'
