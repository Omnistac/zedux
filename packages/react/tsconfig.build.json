{
  "extends": "../../tsconfig.build.json",
  "compilerOptions": {
    "lib": ["DOM", "ESNext"],
    "outDir": "./dist/esm",
    "paths": {
      "@zedux/react": ["./packages/react/src"],
      "@zedux/react/*": ["./packages/react/src/*"]
    }
  },
  "include": ["./src", "../../global.d.ts"]
}
