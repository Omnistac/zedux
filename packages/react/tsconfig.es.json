{
  "compilerOptions": {
    "plugins": [
      {
        "transform": "typescript-transform-paths"
      },
      {
        "transform": "typescript-transform-paths",
        "afterDeclarations": true
      }
    ]
  },
  "extends": "./tsconfig.json",
  "include": ["./src"]
}
