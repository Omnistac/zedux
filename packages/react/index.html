<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Zedux Dev</title>
  </head>
  <body>
    <div id="root"></div>
    <script type="module" src="/src/index.ts"></script>
    <script type="module">
      import { atom, useAtomState } from './src/index'
      import { createElement } from 'react'
      import { render } from 'react-dom'

      const greetingAtom = atom('greeting', 'Hello, World!')

      const App = () => {
        const [state, setState] = useAtomState(greetingAtom)

        return createElement('div', null,
          createElement('p', null, state),
          createElement('input', {
            onChange: event => setState(event.currentTarget.value),
            value: state
          })
        )
      }

      render(createElement(App), document.getElementById('root'))
    </script>
  </body>
</html>
