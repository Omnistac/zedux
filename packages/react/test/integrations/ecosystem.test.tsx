import { render } from '@testing-library/react'
import { atom, ecosystem, injectStore, injectWhy } from '@zedux/react'
import React from 'react'

const testEcosystem = ecosystem({ id: 'test' })

afterEach(() => {
  testEcosystem.wipe()
})

describe('ecosystem', () => {
  test('big graph', async () => {
    const evaluations: number[] = []
    const evaluate1 = jest.fn(() => evaluations.push(1))
    const evaluate2 = jest.fn(() => evaluations.push(2))
    const evaluate3 = jest.fn(() => evaluations.push(3))
    const evaluate4 = jest.fn(() => evaluations.push(4))
    const evaluate5 = jest.fn(() => evaluations.push(5))
    const childRendered = jest.fn()
    const why5 = jest.fn()

    const atom1 = atom('atom1', () => {
      evaluate1()
      const store = injectStore('1', true)

      return store
    })

    const atom2 = atom('atom2', () => {
      evaluate2()
      const atom1val = atom1.injectValue()

      return `${atom1val} 2`
    })

    const atom3 = atom('atom3', (id: string) => {
      evaluate3()
      const atom1val = atom1.injectValue()
      const atom2val = atom2.injectValue()

      return `${id} ${atom1val} ${atom2val}`
    })

    const atom4 = atom('atom4', () => {
      evaluate4()
      const atom3val = atom3.injectValue('1')
      const atom1val = atom1.injectValue()

      return `${atom3val} ${atom1val}`
    })

    const atom5 = atom('atom5', () => {
      evaluate5()
      const atom2val = atom2.injectValue()
      const atom4val = atom4.injectValue()
      const atom1val = atom1.injectValue()

      injectWhy(why5)

      return `${atom4val} ${atom2val} ${atom1val}`
    })

    function Child() {
      const atom5val = atom5.useValue()
      const atom4val = atom4.useValue()
      const atom3val = atom3.useValue('1')
      const atom2val = atom2.useValue()
      const atom1val = atom1.useValue()

      childRendered(atom5val, atom4val, atom3val, atom2val, atom1val)

      return (
        <>
          <div>{atom1val}</div>
          <div>{atom2val}</div>
          <div>{atom3val}</div>
          <div>{atom4val}</div>
          <div>{atom5val}</div>
        </>
      )
    }

    function Test() {
      return (
        <testEcosystem.Provider>
          <Child />
        </testEcosystem.Provider>
      )
    }

    const { findByText } = render(<Test />)

    expect(testEcosystem.instances).toEqual({
      atom1: expect.any(Object),
      atom2: expect.any(Object),
      'atom3-["1"]': expect.any(Object),
      atom4: expect.any(Object),
      atom5: expect.any(Object),
    })
    expect(childRendered).toHaveBeenCalledTimes(1)
    expect(childRendered).toHaveBeenLastCalledWith(
      '1 1 1 2 1 1 2 1',
      '1 1 1 2 1',
      '1 1 1 2',
      '1 2',
      '1'
    )
    expect(evaluations).toEqual([5, 2, 1, 4, 3])

    testEcosystem.load(atom1).setState('0')

    await findByText('1 0 0 2 0 0 2 0')

    expect(childRendered).toHaveBeenCalledTimes(6) // unfortunate but what can be done
    expect(childRendered).toHaveBeenLastCalledWith(
      '1 0 0 2 0 0 2 0',
      '1 0 0 2 0',
      '1 0 0 2',
      '0 2',
      '0'
    )
    expect(evaluations).toEqual([5, 2, 1, 4, 3, 1, 2, 3, 4, 5])
  })
})
