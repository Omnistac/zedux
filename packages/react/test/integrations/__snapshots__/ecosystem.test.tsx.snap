// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ecosystem big graph 1`] = `
{
  "atom1": {
    "dependencies": [],
    "dependents": [
      {
        "key": "atom2",
        "operation": "injectAtomValue",
      },
      {
        "key": "atom3-["1"]",
        "operation": "injectAtomValue",
      },
      {
        "key": "atom4",
        "operation": "injectAtomValue",
      },
      {
        "key": "atom5",
        "operation": "injectAtomValue",
      },
      {
        "key": "Child-4",
        "operation": "useAtomValue",
      },
    ],
    "weight": 1,
  },
  "atom2": {
    "dependencies": [
      {
        "key": "atom1",
        "operation": "injectAtomValue",
      },
    ],
    "dependents": [
      {
        "key": "atom3-["1"]",
        "operation": "injectAtomValue",
      },
      {
        "key": "atom5",
        "operation": "injectAtomValue",
      },
      {
        "key": "Child-3",
        "operation": "useAtomValue",
      },
    ],
    "weight": 2,
  },
  "atom3-["1"]": {
    "dependencies": [
      {
        "key": "atom1",
        "operation": "injectAtomValue",
      },
      {
        "key": "atom2",
        "operation": "injectAtomValue",
      },
    ],
    "dependents": [
      {
        "key": "atom4",
        "operation": "injectAtomValue",
      },
      {
        "key": "Child-2",
        "operation": "useAtomValue",
      },
    ],
    "weight": 4,
  },
  "atom4": {
    "dependencies": [
      {
        "key": "atom3-["1"]",
        "operation": "injectAtomValue",
      },
      {
        "key": "atom1",
        "operation": "injectAtomValue",
      },
    ],
    "dependents": [
      {
        "key": "atom5",
        "operation": "injectAtomValue",
      },
      {
        "key": "Child-1",
        "operation": "useAtomState",
      },
    ],
    "weight": 6,
  },
  "atom5": {
    "dependencies": [
      {
        "key": "atom2",
        "operation": "injectAtomValue",
      },
      {
        "key": "atom4",
        "operation": "injectAtomValue",
      },
      {
        "key": "atom1",
        "operation": "injectAtomValue",
      },
    ],
    "dependents": [
      {
        "key": "Child-0",
        "operation": "useAtomValue",
      },
    ],
    "weight": 10,
  },
}
`;

exports[`ecosystem big graph 2`] = `
{
  "atom5": {
    "atom1": {},
    "atom2": {
      "atom1": {},
    },
    "atom4": {
      "atom1": {},
      "atom3-["1"]": {
        "atom1": {},
        "atom2": {
          "atom1": {},
        },
      },
    },
  },
}
`;

exports[`ecosystem big graph 3`] = `
{
  "atom1": {
    "atom2": {
      "atom3-["1"]": {
        "atom4": {
          "atom5": {},
        },
      },
      "atom5": {},
    },
    "atom3-["1"]": {
      "atom4": {
        "atom5": {},
      },
    },
    "atom4": {
      "atom5": {},
    },
    "atom5": {},
  },
}
`;
